### PR TITLE
SAN-4095 Caching

### PR DESCRIPTION
**Pull request by Myztiq**
Added caching support to help to optimize swarmInfo calls.
- [x] @anandkumarpatel 
- [x] @bkendall

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -37,6 +37,7 @@
   },
   "dependencies": {
     "101": "^1.2.0",
+    "bluebird": "^3.3.5",
```
Comment (anandkumarpatel) [inline package.json:40]: hmm not sure if there was a reason we did not start by using promises
@bkendall ?

File: index.js [hunk 2]
```diff
@@ -3,6 +3,24 @@
...
+function handleCache (key, cacheFetch, cb) {
+
+  // If CACHE_LENGTH is not set we don't want to cache anything
+  if (!exists(process.env.SWARMERODE_CACHE_LENGTH)) {
+    return cacheFetch(cb)
+  }
+
+  if (!cache[key]) {
+    cache[key] = Promise.fromCallback(cacheFetch)
+    setTimeout(function () {
```
Comment (anandkumarpatel) [inline index.js]: why not `.delay()` ?

File: package.json (hunk 1 above)
Comment (Myztiq) [inline package.json:40]: It's just easier to implement caching using promises, that way failures are also cached (which can be considered good in some cases)

File: index.js (hunk 2 above)
Comment (Myztiq) [inline index.js]: `.delay()` only gets called on success. Also we don't want to delay the response. I could do `cacheFetch.catch(noop).delay(TIME).then(`

File: index.js [hunk 3]
```diff
@@ -3,6 +3,26 @@
...
+
+function handleCache (key, cacheFetch, cb) {
+  // If SWARMERODE_CACHE_LENGTH is not set we don't want to cache anything
+  if (!exists(process.env.SWARMERODE_CACHE_LENGTH)) {
+    return cacheFetch(cb)
+  }
+
+  if (!cache[key]) {
+    cache[key] = Promise.fromCallback(cacheFetch)
+    cache[key]
```
Comment (anandkumarpatel) [inline index.js]: why not just `Promise.resolve().delay()

Comment (Myztiq) [inline index.js]: we want to wait for the cacheFetch to actually resolve before starting the timer.

Comment (bkendall) [inline index.js]: it's a promise tree.

File: test/index.js [hunk 4]
```diff
@@ -120,6 +120,49 @@ describe('Swarmerode', function () {
...
+          sinon.assert.calledTwice(Swarmerode._Swarmerode._parseSwarmSystemStatus)
+          sinon.assert.calledWith(Swarmerode._Swarmerode._parseSwarmSystemStatus, 1)
+          sinon.assert.calledWith(Swarmerode._Swarmerode._parseSwarmSystemStatus, 2)
+          done()
+        }
+      }
+      instance.swarmInfo(handleCb)
+      instance.swarmInfo(handleCb)
+    })
+  })
```
Comment (bkendall) [inline test/index.js]: follow the pattern in this file - blocks are padded

File: index.js [hunk 5]
```diff
@@ -3,6 +3,26 @@
 var clone = require('101/clone')
 var debug = require('debug')('swarmerode')
 var exists = require('101/exists')
+var Promise = require('bluebird')
+var cache = {}
+
+function handleCache (key, cacheFetch, cb) {
```
Comment (bkendall) [inline index.js]: document this function. I'd actually consider moving it to it's own file as well, which would make it easy to unit test

File: package.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "swarmerode",
-  "version": "2.1.0",
+  "version": "2.0.0",
```
Comment (bkendall) [inline package.json]: how did you manage this?

File: test/index.js [hunk 7]
```diff
@@ -156,6 +157,21 @@ describe('Swarmerode', function () {
     })
   })
 
+  describe('swarmInfo', function () {
```
Comment (bkendall) [inline test/index.js]: you've duplicated the `swarmInfo` describe block

File: package.json (hunk 6 above)
Comment (Myztiq) [inline package.json]: shitty merge!

**Comment by bkendall** [comment]
Last thing: can you document this new functionality in the README?

**Comment by bkendall** [comment]
LGTM

**Comment by anandkumarpatel** [comment]
LGTM

**Pull request merged**